app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php

```diff
@@ -65,11 +117,46 @@ public function execute(\Magento\Framework\Event\Observer $observer)
             || $product->getIsChangedWebsites()
             || $product->dataHasChangedFor('visibility')
         ) {
-            if ($product->isVisibleInSiteVisibility()) {
-                $product->unsUrlPath();
-                $product->setUrlPath($this->productUrlPathGenerator->getUrlPath($product));
+            //Refresh rewrite urls
+            $product->unsUrlPath();
+            $product->setUrlPath($this->productUrlPathGenerator->getUrlPath($product));
+            if (!empty($this->productUrlRewriteGenerator->generate($product))) {
                 $this->urlPersist->replace($this->productUrlRewriteGenerator->generate($product));
             }
+
+            $storeIdsToRemove = [];
+            if ($this->productScopeRewriteGenerator->isGlobalScope($product->getStoreId())) {
+                //Remove any rewrite URLs for websites the product is not in, or is not visible in. Global Scope.
+                foreach ($this->storeManager->getWebsites() as $website) {
+                    $websiteId = $website->getWebsiteId();
+                    foreach ($this->storeWebsiteRelation->getStoreByWebsiteId($websiteId) as $storeid) {
+                        //Load the product for the store we are processing so we can see if it is visible
+                        $storeProduct = $this->productRepository->getById(
+                            $product->getId(),
+                            false,
+                            $storeid,
+                            true
+                        );
+                        if (!$storeProduct->isVisibleInSiteVisibility() ||
+                            !in_array($websiteId, $product->getWebsiteIds())) {
```

Comment:
In array in loop is not a good thing from performance. I know that loading product will be much more noticable, but anyway. We can create map outsize of the loop, like this `$productWebsiteMap = array_flip($product->getWebsiteIds());`
and in the loop check `isset($productWebsiteMap[$websiteId])`. It will be really faster